### PR TITLE
Distill

### PR DESCRIPTION
**Pull request by unconst**
Adds distillation into the mnist main synapse. Leaves all other files unchanged.
Distillation minorly reduces the performance in this commit for MNIST which is to be suspected. 
Also, adds back the local model params with the optimizer in main. This creates a double gradient update over the synapse model. 
Not optimal, but will be changed in later commits.

### CODE REVIEW

File: examples/mnist/main.py [hunk 1]
```diff
@@ -48,23 +57,33 @@ def outdef(self):
...
+        x = x.view(-1, 1, 28, 28)
+        x = torch.tanh(self.dist_conv1(x))
+        x = self.dist_average1(x)
+        x = torch.tanh(self.dist_conv2(x))
+        x = self.dist_average2(x)
+        x = torch.tanh(self.dist_conv3(x))
+        x = x.view(-1, x.shape[1])
+        x = F.relu(self.dist_fc1(x))
+        x = F.relu(self.dist_fc2(x))
+        return x
```
Comment (shibshib) [inline examples/mnist/main.py:70]: Is the distill always going to look like the fwd pass?

**Review by shibshib — APPROVED** [review]

File: examples/mnist/main.py (hunk 1 above)
Comment (unconst) [reply examples/mnist/main.py:70]: Not always, but in this case the output of the network is [batch_size, 10] to fit mnist. In all likelihood, outputs would be 1024 + embeddings in which case, the distillation layer would learn x -> 1024
